tests/models/test_cachedfeed.py

```diff
@@ -17,162 +18,540 @@
+        tomorrow_result = m(
+            self._db, wl, facets, pagination, tomorrow_vs_now, 0
+        )
+        eq_(tomorrow_result, result)
+        eq_("Someone in the background set tomorrow's content.",
+            tomorrow_result.content)
+        eq_(tomorrow_result.timestamp, tomorrow)
+
+        # Here, the other thread sets .timestamp to a date in the past, and
+        # it loses out to the (apparently) newer feed.
+        def yesterday_vs_now():
+            result.content = "Someone in the background set yesterday's content."
+            result.timestamp = yesterday
+            return "Today's content is fresher."
+        now_result = m(
+            self._db, wl, facets, pagination, yesterday_vs_now, 0
+        )
+
+        # We got the same CachedFeed we've been getting this whole
+        # time, but the outdated data set by the 'background thread'
+        # has been fixed.
+        eq_(result, now_result)
+        eq_("Today's content is fresher.", result.content)
+        assert result.timestamp > yesterday
+
+        # This shouldn't happen, but if the CachedFeed's timestamp or
+        # content are *cleared out* in the background, between the
+        # time the CacheFeed is fetched and the time the refresher
+        # finishes, then we don't know what's going on and we don't
+        # take chances. We create a whole new CachedFeed object for
+        # the updated version of the feed.
+
+        # First, try the situation where .timestamp is cleared out in
+        # the background.
+        def timestamp_cleared_in_background():
+            result.content = "Someone else sets content and clears timestamp."
+            result.timestamp = None
+
+            return "Non-weird content."
+        result2 = m(
+            self._db, wl, facets, pagination, timestamp_cleared_in_background,
+            0
         )
-        assert feed2 != feed
-        eq_(pagination.query_string, feed2.pagination)
-        eq_(facets.query_string, feed2.facets)
+        now = datetime.datetime.utcnow()
+
+        # result2 is a brand new CachedFeed.
+        assert result2 != result
+        eq_("Non-weird content.", result2.content)
+        assert (now - result2.timestamp).total_seconds() < 2
+
+        # We let the background process do whatever it wants to do
+        # with the old one.
+        eq_("Someone else sets content and clears timestamp.", result.content)
+        eq_(None, result.timestamp)
+
+        # Next, test the situation where .content is cleared out.
+        def content_cleared_in_background():
+            result2.content = None
+            result2.timestamp = tomorrow
+
+            return "Non-weird content."
+        result3 = m(
+            self._db, wl, facets, pagination, content_cleared_in_background, 0
+        )
+        now = datetime.datetime.utcnow()
+
+        # Again, a brand new CachedFeed.
+        assert result3 != result2
+        assert result3 != result
```

Comment:
The older CacheFeeds get deleted in the background eventually, right? Where is that happening?

Reply:
CachedFeedReaper in core/monitor.py removes all feeds older than 30 days.